### PR TITLE
Add vxlan support api changes

### PR DESCRIPTION
**Pull request by qinqon**
This depends on https://github.com/nmstate/nmstate/pull/508

Signed-off-by: Quique Llorente <ellorent@redhat.com>

### CODE REVIEW

**Comment by EdDev** [comment]
Looks good to me. We can take it when the nmstate change is merged.

File: devel/api.md [hunk 1]
```diff
@@ -49,6 +49,11 @@
 * [VLAN](#vlan)
     * [`VLAN.ID`](#vlanid)
     * [`VLAN.BASE_IFACE`](#vlanbase_iface)
+* [VXLAN](#vxlan)
+    * [`VXLAN.ID`](#vxlanvni)
```
Comment (phoracek) [inline devel/api.md]: isn't there a bug in the link?

**Review by phoracek — CHANGES_REQUESTED** [review]

File: devel/api.md (hunk 1 above)
Comment (tyll) [reply devel/api.md]: yes, there is.

Comment (qinqon) [reply devel/api.md]: fixed

File: devel/api.md [hunk 2]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+            VXLAN.BASE_IFACE: 'eth1'
+            VXLAN.REMOTE: '192.168.3.3'
+        }
+    ]
+}
+```
+
+## `VXLAN.ID`
+
+Type: `integer`
```
Comment (tyll) [inline devel/api.md:688]: The python class is actually `int`, but it seems the other examples are also using `integer`. Any objections to use `int` everywhere @EdDev @cathay4t - should happen in a separate PR, of course.

File: devel/api.md [hunk 3]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+
+The VXLAN ID.
+
+To change VXLAN ID, please use `InterfaceState.ABSENT` for existing one and add
+new one after that.
+
+
+## `VXLAN.BASE_IFACE`
+
+Type: `string`
```
Comment (tyll) [inline devel/api.md:698]: As with `integer`, it should be `str` IMHO.

File: devel/api.md [hunk 4]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+}
+```
+
+## `VXLAN.ID`
+
+Type: `integer`
+
+The VXLAN ID.
+
+To change VXLAN ID, please use `InterfaceState.ABSENT` for existing one and add
```
Comment (tyll) [inline devel/api.md]: Proposal: To change *the* VXLAN ID,  please remove the interface first by setting the state to `InterfaceState.ABSENT` and then add a new one.

File: devel/api.md [hunk 5]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+new one after that.
+
+
+## `VXLAN.BASE_IFACE`
+
+Type: `string`
+
+The interface name for VXLAN underlay.
```
Comment (tyll) [inline devel/api.md]: Proposal: The name of the underlying interface for the VXLAN.

**Review by EdDev — DISMISSED** [review]
Thank you!

File: devel/api.md [hunk 6]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+
+
+## `VXLAN.REMOTE`
+
+Type: `string`
```
Comment (tyll) [inline devel/api.md:704]: `str`

File: devel/api.md [hunk 7]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+
+Specifies the unicast destination IP address to use in outgoing packets when the destination link layer address is not known in the VXLAN device forwarding database, or the multicast IP address to join.
+
+## `VXLAN.DESTINATION_PORT`
+
+Type: `integer`
```
Comment (tyll) [inline devel/api.md:710]: `int`

File: devel/api.md [hunk 8]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+Type: `string`
+
+The interface name for VXLAN underlay.
+
+
+## `VXLAN.REMOTE`
+
+Type: `string`
+
+Specifies the unicast destination IP address to use in outgoing packets when the destination link layer address is not known in the VXLAN device forwarding database, or the multicast IP address to join.
```
Comment (tyll) [inline devel/api.md]: To make it similar to the other entries, the sentence should not start with `Specifies` but with `The unicast destination address...`

File: devel/api.md [hunk 9]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+
+## `VXLAN.DESTINATION_PORT`
+
+Type: `integer`
+
+Specifies the UDP destination port to communicate to the remote VXLAN tunnel endpoint.
```
Comment (tyll) [inline devel/api.md]: Proposal: The UDP destination port of the remote VXLAN tunnel endpoint.

Comment (tyll) [reply devel/api.md]: Does this default to `4789`? Should the default be mentioned?

File: devel/api.md [hunk 10]
```diff
@@ -655,6 +661,57 @@ Type: `string`
...
+{
+    Interface.KEY: [
+        Interface.NAME: "eth1.101",
+        Interface.TYPE: VXLAN.TYPE,
+        Interface.STATE: InterfaceState.UP
+        VXLAN.CONFIG_SUBTREE: {
+            VXLAN.ID: 101,
+            VXLAN.BASE_IFACE: 'eth1'
+            VXLAN.REMOTE: '192.168.3.3'
+        }
```
Comment (tyll) [inline devel/api.md:681]: How about also adding the destination port and adding a comment `# optional`.

File: devel/api.md (hunk 2 above)
Comment (qinqon) [reply devel/api.md:688]: I am ok with another PR, but not here.

File: devel/api.md (hunk 3 above)
Comment (qinqon) [reply devel/api.md:698]: ditto

File: devel/api.md [hunk 11]
```diff
@@ -655,6 +661,58 @@ Type: `string`
...
+```python
+{
+    Interface.KEY: [
+        Interface.NAME: "eth1.101",
+        Interface.TYPE: VXLAN.TYPE,
+        Interface.STATE: InterfaceState.UP
+        VXLAN.CONFIG_SUBTREE: {
+            VXLAN.ID: 101,
+            VXLAN.BASE_IFACE: 'eth1'
+            VXLAN.REMOTE: '192.168.3.3'
```
Comment (tyll) [inline devel/api.md]: There are commas missing in the example. Can you maybe pipe it through black to see that it is correct? Also the string constants use different quote notations.

File: devel/api.md [hunk 12]
```diff
@@ -655,6 +661,58 @@ Type: `string`
...
+Type: `string`
+
+The unicast destination IP address to use in outgoing packets when the destination link layer address is not known in the VXLAN device forwarding database, or the multicast IP address to join.
+
+## `VXLAN.DESTINATION_PORT`
+
+Type: `integer`
+
+The UDP destination port to communicate to the remote VXLAN tunnel endpoint.
+default: 4789
```
Comment (tyll) [inline devel/api.md]: The default should be upper-case now.

File: devel/api.md (hunk 11 above)
Comment (tyll) [reply devel/api.md]: Also please use RFC5737 IP addresses: https://tools.ietf.org/html/rfc5737

Comment (qinqon) [reply devel/api.md]: Didn't know about it, that's cool.

File: devel/api.md [hunk 13]
```diff
@@ -655,6 +661,58 @@ Type: `string`
...
+# VXLAN
+Besides basic interface properties, each VXLAN interface state also contains
+a dictionary saved in key `VXLAN.CONFIG_SUBTREE`.
+
+Example:
+
+```python
+{
+    Interface.KEY: [{
+        Interface.NAME: "eth1.101",
```
Comment (tyll) [inline devel/api.md]: This line is using double quotes but the other lines are using single quotes, please use a uniform style.

**Review by tyll — APPROVED** [review]
Thank you, LGTM!

**Pull request merged**